include/zephyr/rtio/rtio.h

```diff
@@ -536,22 +578,25 @@ static inline void rtio_sqe_prep_transceive(struct rtio_sqe *sqe,
  * @param sq_sz Size of the submission queue, must be power of 2
  * @param cq_sz Size of the completion queue, must be power of 2
  */
-#define RTIO_DEFINE(name, exec, sq_sz, cq_sz)	\
-	IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM,							   \
-		   (static K_SEM_DEFINE(_submit_sem_##name, 0, K_SEM_MAX_LIMIT)))		   \
-	IF_ENABLED(CONFIG_RTIO_CONSUME_SEM,							   \
-		   (static K_SEM_DEFINE(_consume_sem_##name, 0, K_SEM_MAX_LIMIT)))		   \
-	static RTIO_SQ_DEFINE(_sq_##name, sq_sz);						   \
-	static RTIO_CQ_DEFINE(_cq_##name, cq_sz);						   \
-	STRUCT_SECTION_ITERABLE(rtio, name) = {							   \
-		.executor = (exec),                                                                \
-		.xcqcnt = ATOMIC_INIT(0),                                                          \
-		IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM, (.submit_sem = &_submit_sem_##name,))	   \
-		IF_ENABLED(CONFIG_RTIO_SUBMIT_SEM, (.submit_count = 0,))			   \
-		IF_ENABLED(CONFIG_RTIO_CONSUME_SEM, (.consume_sem = &_consume_sem_##name,))	   \
-		.sq = (struct rtio_sq *const)&_sq_##name,					   \
-		.cq = (struct rtio_cq *const)&_cq_##name,                                          \
-	};
+#define RTIO_DEFINE(name, exec, sq_sz, cq_sz) _RTIO_DEFINE(name, exec, sq_sz, cq_sz, NULL, 0)
+
+/**
+ * @brief Statically define and initialize an RTIO context
+ *
+ * @param name Name of the RTIO
+ * @param exec Symbol for rtio_executor (pointer)
+ * @param sq_sz Size of the submission queue, must be power of 2
+ * @param cq_sz Size of the completion queue, must be power of 2
+ * @param num_blks Number of blocks in the memory pool
+ * @param blk_size The number of bytes in each block
+ * @param balign The block aligmnent
+ */
+#ifdef CONFIG_RTIO_SYS_MEM_BLOCKS
+#define RTIO_DEFINE_WITH_MEMPOOL(name, exec, sq_sz, cq_sz, num_blks, blk_size, balign)             \
+	SYS_MEM_BLOCKS_DEFINE_STATIC(_mempool_##name, blk_size, num_blks, balign);                 \
```

Comment:
You'll need to create a memory partition and domain for this to work out the way you are looking for in user mode, but its not far off.

You'll need to define a partition, add it to a domain, then add a thread to that domain...
E.g.

right here define the partition (you may want to make it one partition and domain where all rtio mem pools are at, there's limited numbers of domains typically for arm mpus):

K_MEM_PARTITION_DEFINE(_partition_##name, &mem_pool_buf[0], &mem_pool_buf[blk_size*num_blks], K_MEM_PARTITION_P_RW_U_RO);

From there, you need to have a k_mem_domain member of struct rtio (or reference one), init it

In some rtio_init function run once somewhere...

k_mem_domain_init(&r->domain, 0, NULL);
k_mem_domain_add_partition(&r->domain, r->mempool_buf_partition);

Finally in rtio_access_grant

k_mem_domain_add_thread(&r->domain, tid);

Reply:
Hmmm I'm trying to think how to make the mempools for all RTIO context belong in a single partition and this seems complex. I guess we would have to create an iterable section for it, then where we initialize the domain assign a start for the mempool for each rtio? Lets revisit that optimization on the next iteration?

Reply:
Agreed